File: javaspider/src/main/java/com/javaspider/selector/DevToolsAnalyzer.java
package com.javaspider.selector;

import org.jsoup.Jsoup;
import org.jsoup.nodes.Document;
import org.jsoup.nodes.Element;

import java.util.ArrayList;
import java.util.Comparator;
import java.util.LinkedHashMap;
import java.util.List;
import java.util.Locale;
import java.util.Map;

public final class DevToolsAnalyzer {
    private static final List<String> IMPORTANT_ATTRS = List.of(
        "id", "class", "name", "type", "href", "src", "role", "aria-label",
        "data-testid", "data-test", "placeholder", "action", "method"
    );

    public DevToolsReport analyze(String html, List<DevToolsNetworkArtifact> network, List<Map<String, String>> consoleEvents) {
        Document document = Jsoup.parse(html == null ? "" : html);
        List<ElementSnapshot> elements = new ArrayList<>();
        for (Element element : document.getAllElements()) {
            if ("#root".equals(element.tagName())) {
                continue;
            }
            elements.add(snapshot(element));
        }
        List<String> sources = new ArrayList<>();
        List<String> inline = new ArrayList<>();
        for (Element script : document.select("script")) {
            String src = script.attr("src").trim();
            if (!src.isBlank()) {
                sources.add(src);
                continue;
            }
            String code = script.data().isBlank() ? script.html().trim() : script.data().trim();
            if (!code.isBlank()) {
                inline.add(code.length() > 2000 ? code.substring(0, 2000) : code);
            }
        }
        List<DevToolsNetworkArtifact> networkCandidates = networkCandidates(network == null ? List.of() : network);
        List<ReverseRecommendation> recommendations = recommendReverseRoutes(html == null ? "" : html, sources, inline, networkCandidates);
        Map<String, Object> summary = new LinkedHashMap<>();
        summary.put("element_count", elements.size());
        summary.put("script_count", sources.size() + inline.size());
        summary.put("network_candidate_count", networkCandidates.size());
        summary.put("best_reverse_route", recommendations.isEmpty() ? "" : recommendations.get(0).kind());
        return new DevToolsReport(
            elements,
            sources,
            inline,
            networkCandidates,
            consoleEvents == null ? List.of() : consoleEvents,
            recommendations,
            summary
        );
    }

    private ElementSnapshot snapshot(Element element) {
        Map<String, String> attrs = new LinkedHashMap<>();
        for (String attr : IMPORTANT_ATTRS) {
            String value = element.attr(attr).trim();
            if (!value.isBlank()) {
                attrs.put(attr, value);
            }
        }
        String text = element.text().trim();
        if (text.length() > 120) {
            text = text.substring(0, 120);
        }
        return new ElementSnapshot(element.tagName(), element.cssSelector(), fullXPath(element), text, attrs);
    }

    private List<DevToolsNetworkArtifact> networkCandidates(List<DevToolsNetworkArtifact> network) {
        List<DevToolsNetworkArtifact> result = new ArrayList<>();
        java.util.Set<String> seen = new java.util.LinkedHashSet<>();
        for (DevToolsNetworkArtifact entry : network) {
            if (entry == null || entry.url() == null || entry.url().isBlank() || seen.contains(entry.url())) {
                continue;
            }
            String type = entry.resourceType() == null ? "" : entry.resourceType().toLowerCase(Locale.ROOT);
            String signal = (entry.url() + " " + type).toLowerCase(Locale.ROOT);
            if (List.of("script", "xhr", "fetch", "websocket", "document").contains(type)
                || List.of("api", "sign", "token", "encrypt", "decrypt", "jsonp", "webpack").stream().anyMatch(signal::contains)) {
                seen.add(entry.url());
                result.add(entry);
            }
        }
        return result;
    }

    private List<ReverseRecommendation> recommendReverseRoutes(
        String html,
        List<String> sources,
        List<String> inline,
        List<DevToolsNetworkArtifact> network
    ) {
        StringBuilder corpusBuilder = new StringBuilder(html.length() > 8000 ? html.substring(0, 8000) : html);
        sources.forEach(value -> corpusBuilder.append('\n').append(value));
        inline.forEach(value -> corpusBuilder.append('\n').append(value));
        network.forEach(value -> corpusBuilder.append('\n').append(value.url()));
        String corpus = corpusBuilder.toString().toLowerCase(Locale.ROOT);

        List<ReverseRecommendation> result = new ArrayList<>();
        addRecommendation(result, corpus, "analyze_crypto", 100, "发现加密、签名或摘要相关标记，优先交给 Node.js crypto 逆向分析",
            List.of("cryptojs", "crypto.subtle", "aes", "rsa", "md5", "sha1", "sha256", "encrypt", "decrypt", "signature", "sign"));
        addRecommendation(result, corpus, "analyze_webpack", 90, "发现 webpack 模块运行时，适合进入模块表和导出函数逆向",
            List.of("__webpack_require__", "webpackjsonp", "webpackchunk", "webpack://"));
        addRecommendation(result, corpus, "simulate_browser", 80, "脚本依赖浏览器运行时对象，适合用 Node.js 浏览器环境模拟",
            List.of("localstorage", "sessionstorage", "navigator.", "document.", "window.", "canvas", "webdriver"));
        addRecommendation(result, corpus, "analyze_ast", 60, "存在外链或内联脚本，适合进行 AST 结构分析和函数定位",
            List.of(".js", "function", "=>", "eval(", "new function"));
        result.sort(Comparator.comparingInt(ReverseRecommendation::priority).reversed().thenComparing(ReverseRecommendation::kind));
        return result;
    }

    private void addRecommendation(List<ReverseRecommendation> result, String corpus, String kind, int priority, String reason, List<String> markers) {
        List<String> evidence = markers.stream()
            .filter(marker -> corpus.contains(marker.toLowerCase(Locale.ROOT)))
            .limit(8)
            .toList();
        if (!evidence.isEmpty()) {
            result.add(new ReverseRecommendation(kind, priority, reason, evidence));
        }
    }

    private String fullXPath(Element element) {
        List<String> parts = new ArrayList<>();
        Element current = element;
        while (current != null && !"#root".equals(current.tagName())) {
            if (!current.id().isBlank()) {
                parts.add(current.tagName() + "[@id=" + xpathLiteral(current.id()) + "]");
                break;
            }
            int index = 1;
            Element sibling = current.previousElementSibling();
            while (sibling != null) {
                if (sibling.tagName().equals(current.tagName())) {
                    index++;
                }
                sibling = sibling.previousElementSibling();
            }
            parts.add(current.tagName() + "[" + index + "]");
            current = current.parent();
        }
        java.util.Collections.reverse(parts);
        return parts.isEmpty() ? "" : "/" + String.join("/", parts);
    }

    private String xpathLiteral(String value) {
        if (!value.contains("'")) {
            return "'" + value + "'";
        }
        if (!value.contains("\"")) {
            return "\"" + value + "\"";
        }
        String[] parts = value.split("'");
        return "concat(" + String.join(", \"'\", ", java.util.Arrays.stream(parts).map(part -> "'" + part + "'").toList()) + ")";
    }

    public record ElementSnapshot(String tag, String css, String xpath, String text, Map<String, String> attrs) {
    }

    public record DevToolsNetworkArtifact(String url, String method, int status, String resourceType) {
    }

    public record ReverseRecommendation(String kind, int priority, String reason, List<String> evidence) {
    }

    public record DevToolsReport(
        List<ElementSnapshot> elements,
        List<String> scriptSources,
        List<String> inlineScriptSamples,
        List<DevToolsNetworkArtifact> networkCandidates,
        List<Map<String, String>> consoleEvents,
        List<ReverseRecommendation> reverseRecommendations,
        Map<String, Object> summary
    ) {
        public ReverseRecommendation bestReverseRoute() {
            return reverseRecommendations.isEmpty() ? null : reverseRecommendations.get(0);
        }
    }
}

File: gospider/parser/devtools_analyzer.go
package parser

import (
	"net/url"
	"sort"
	"strings"

	"github.com/PuerkitoBio/goquery"
	"golang.org/x/net/html"
)

type ElementSnapshot struct {
	Tag   string            `json:"tag"`
	CSS   string            `json:"css"`
	XPath string            `json:"xpath"`
	Text  string            `json:"text,omitempty"`
	Attrs map[string]string `json:"attrs,omitempty"`
}

type DevToolsNetworkArtifact struct {
	URL          string `json:"url"`
	Method       string `json:"method,omitempty"`
	Status       int    `json:"status,omitempty"`
	ResourceType string `json:"resource_type,omitempty"`
}

type ReverseRecommendation struct {
	Kind     string   `json:"kind"`
	Priority int      `json:"priority"`
	Reason   string   `json:"reason"`
	Evidence []string `json:"evidence"`
}

type DevToolsReport struct {
	Elements               []ElementSnapshot         `json:"elements"`
	ScriptSources          []string                  `json:"script_sources"`
	InlineScriptSamples    []string                  `json:"inline_script_samples"`
	NetworkCandidates      []DevToolsNetworkArtifact `json:"network_candidates"`
	ConsoleEvents          []map[string]string       `json:"console_events,omitempty"`
	ReverseRecommendations []ReverseRecommendation   `json:"reverse_recommendations"`
	Summary                map[string]interface{}    `json:"summary"`
}

func (r *DevToolsReport) BestReverseRoute() *ReverseRecommendation {
	if r == nil || len(r.ReverseRecommendations) == 0 {
		return nil
	}
	return &r.ReverseRecommendations[0]
}

type DevToolsAnalyzer struct{}

func NewDevToolsAnalyzer() *DevToolsAnalyzer {
	return &DevToolsAnalyzer{}
}

func (a *DevToolsAnalyzer) Analyze(htmlText string, network []DevToolsNetworkArtifact, console []map[string]string) (*DevToolsReport, error) {
	parser := NewHTMLParser(htmlText)
	if parser == nil || parser.doc == nil {
		return nil, nil
	}
	elements := make([]ElementSnapshot, 0)
	parser.doc.Find("*").Each(func(_ int, selection *goquery.Selection) {
		if node := selection.Get(0); node != nil {
			elements = append(elements, snapshotElement(selection, node))
		}
	})
	sources, inline := scriptArtifacts(parser.doc)
	networkCandidates := networkCandidates(network)
	recommendations := recommendReverseRoutes(htmlText, sources, inline, networkCandidates)
	return &DevToolsReport{
		Elements:               elements,
		ScriptSources:          sources,
		InlineScriptSamples:    inline,
		NetworkCandidates:      networkCandidates,
		ConsoleEvents:          console,
		ReverseRecommendations: recommendations,
		Summary: map[string]interface{}{
			"element_count":           len(elements),
			"script_count":            len(sources) + len(inline),
			"network_candidate_count": len(networkCandidates),
			"best_reverse_route":      bestReverseKind(recommendations),
		},
	}, nil
}

func snapshotElement(selection *goquery.Selection, node *html.Node) ElementSnapshot {
	attrs := map[string]string{}
	for _, attr := range []string{"id", "class", "name", "type", "href", "src", "role", "aria-label", "data-testid", "data-test", "placeholder", "action", "method"} {
		if value := strings.TrimSpace(attrValue(node, attr)); value != "" {
			attrs[attr] = value
		}
	}
	text := strings.TrimSpace(selection.Text())
	if len([]rune(text)) > 120 {
		text = string([]rune(text)[:120])
	}
	return ElementSnapshot{
		Tag:   goquery.NodeName(selection),
		CSS:   fullCSSPath(node),
		XPath: fullXPath(node),
		Text:  text,
		Attrs: attrs,
	}
}

func scriptArtifacts(doc *goquery.Document) ([]string, []string) {
	sources := []string{}
	inline := []string{}
	doc.Find("script").Each(func(_ int, selection *goquery.Selection) {
		if src, ok := selection.Attr("src"); ok && strings.TrimSpace(src) != "" {
			sources = append(sources, strings.TrimSpace(src))
			return
		}
		code := strings.TrimSpace(selection.Text())
		if code == "" {
			return
		}
		if len([]rune(code)) > 2000 {
			code = string([]rune(code)[:2000])
		}
		inline = append(inline, code)
	})
	return sources, inline
}

func networkCandidates(network []DevToolsNetworkArtifact) []DevToolsNetworkArtifact {
	result := []DevToolsNetworkArtifact{}
	seen := map[string]bool{}
	for _, entry := range network {
		if entry.URL == "" || seen[entry.URL] {
			continue
		}
		signal := strings.ToLower(entry.URL + " " + entry.ResourceType)
		if isReverseNetworkCandidate(signal, strings.ToLower(entry.ResourceType)) {
			seen[entry.URL] = true
			result = append(result, normalizeNetworkArtifact(entry))
		}
	}
	return result
}

func normalizeNetworkArtifact(entry DevToolsNetworkArtifact) DevToolsNetworkArtifact {
	if parsed, err := url.Parse(entry.URL); err == nil && parsed.Host != "" {
		entry.URL = parsed.String()
	}
	return entry
}

func isReverseNetworkCandidate(signal, resourceType string) bool {
	if resourceType == "script" || resourceType == "xhr" || resourceType == "fetch" || resourceType == "websocket" || resourceType == "document" {
		return true
	}
	for _, token := range []string{"api", "sign", "token", "encrypt", "decrypt", "jsonp", "webpack"} {
		if strings.Contains(signal, token) {
			return true
		}
	}
	return false
}

func recommendReverseRoutes(htmlText string, sources []string, inline []string, network []DevToolsNetworkArtifact) []ReverseRecommendation {
	parts := []string{htmlText}
	parts = append(parts, sources...)
	parts = append(parts, inline...)
	for _, entry := range network {
		parts = append(parts, entry.URL)
	}
	corpus := strings.ToLower(strings.Join(parts, "\n"))
	recommendations := []ReverseRecommendation{}
	for _, spec := range []struct {
		kind     string
		priority int
		reason   string
		markers  []string
	}{
		{"analyze_crypto", 100, "发现加密、签名或摘要相关标记，优先交给 Node.js crypto 逆向分析", []string{"cryptojs", "crypto.subtle", "aes", "rsa", "md5", "sha1", "sha256", "encrypt", "decrypt", "signature", "sign"}},
		{"analyze_webpack", 90, "发现 webpack 模块运行时，适合进入模块表和导出函数逆向", []string{"__webpack_require__", "webpackjsonp", "webpackchunk", "webpack://"}},
		{"simulate_browser", 80, "脚本依赖浏览器运行时对象，适合用 Node.js 浏览器环境模拟", []string{"localstorage", "sessionstorage", "navigator.", "document.", "window.", "canvas", "webdriver"}},
		{"analyze_ast", 60, "存在外链或内联脚本，适合进行 AST 结构分析和函数定位", []string{".js", "function", "=>", "eval(", "new function"}},
	} {
		evidence := matchedMarkers(corpus, spec.markers)
		if len(evidence) == 0 {
			continue
		}
		recommendations = append(recommendations, ReverseRecommendation{Kind: spec.kind, Priority: spec.priority, Reason: spec.reason, Evidence: evidence})
	}
	sort.Slice(recommendations, func(i, j int) bool {
		if recommendations[i].Priority != recommendations[j].Priority {
			return recommendations[i].Priority > recommendations[j].Priority
		}
		return recommendations[i].Kind < recommendations[j].Kind
	})
	return recommendations
}

func matchedMarkers(corpus string, markers []string) []string {
	result := []string{}
	for _, marker := range markers {
		if strings.Contains(corpus, strings.ToLower(marker)) {
			result = append(result, marker)
		}
		if len(result) >= 8 {
			break
		}
	}
	return result
}

func bestReverseKind(recommendations []ReverseRecommendation) string {
	if len(recommendations) == 0 {
		return ""
	}
	return recommendations[0].Kind
}
